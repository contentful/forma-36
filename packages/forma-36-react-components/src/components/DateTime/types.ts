export type ISO8601Timestamp = string;
export type UNIXEpochMillis = number;
/**
 * CoercibleDate is a date represented as a JS Date object or something that can be coerced to a
 * JS Date object.
 *
 * @example new Date()
 * @example '2020-08-12T01:23:45.000Z'
 * @example 1597195425000
 **/
export type CoercibleDate = Date | ISO8601Timestamp | UNIXEpochMillis;
