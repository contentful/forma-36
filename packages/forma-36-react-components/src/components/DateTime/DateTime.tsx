import React from 'react';

import {
  DateTimeFormat,
  formatDateTime,
  formatMachineReadableDateTime,
} from './dateUtils';
import { CoercibleDate } from './types';

export interface DateTimeProps {
  /**
   * The date to display as a JS Date, an ISO8601 Timestamp, or Unix Epoch Milliseconds
   *
   * @example '2020-08-13T01:23:45.000Z'
   * @example new Date()
   * @example Date.now()
   */
  date: CoercibleDate;
  /**
   * Which display format to use
   *
   * @default 'FULL'
   **/
  format?: DateTimeFormat;
  className?: string;
  testId?: string;
}

/**
 * Formats a JS Date object to an absolute date and time inside of a <time /> tag
 */
export const DateTime: React.FC<DateTimeProps> = ({
  date,
  format = 'FULL',
  className,
  testId,
}) => {
  if (typeof date === 'string' || typeof date === 'number') {
    date = new Date(date);
  }
  const formatted = formatDateTime(date, format);
  const machineReadable = formatMachineReadableDateTime(date, format);

  return (
    <time
      dateTime={machineReadable}
      className={className}
      data-test-id={testId}
    >
      {formatted}
    </time>
  );
};
