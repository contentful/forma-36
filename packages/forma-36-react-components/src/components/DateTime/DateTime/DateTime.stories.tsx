import React from 'react';

import DateTime, { DateTimeProps } from './DateTime';

import notes from '../README.md';

const exampleDate = '2020-08-13T13:45:56.0123Z';

export default {
  title: 'Components/DateTime/DateTime',
  component: DateTime,
  parameters: {
    propTypes: DateTime['__docgenInfo'],
    notes,
  },
  argTypes: {
    date: { control: { type: 'date' } },
    format: {
      control: {
        type: 'select',
        options: ['FULL', 'DATE_ONLY', 'TIME_ONLY', 'WEEKDAY_DATE'],
      },
    },
    className: { control: { type: 'text' } },
    testId: { control: { type: 'text' } },
  },
};

const DateTimeStory = (args: DateTimeProps) => <DateTime {...args} />;

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const basic: any = DateTimeStory.bind({});
basic.args = {
  date: exampleDate,
};

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const withOtherFormats: any = DateTimeStory.bind({});
withOtherFormats.args = {
  date: exampleDate,
  format: 'DATE_ONLY',
};
