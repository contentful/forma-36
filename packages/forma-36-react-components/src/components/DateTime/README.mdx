---
title: 'DateTime'
type: 'component'
status: 'stable'
slug: /components/date-time/
github: 'https://github.com/contentful/forma-36/tree/master/packages/forma-36-react-components/src/components/DateTime'
storybook: 'https://f36-storybook.contentful.com/?path=/story/components-datetime-datetime--basic'
---

## Dates and Times

Provides components for displaying dates and times while following the [Forma 36 guidelines](https://f36.contentful.com/guidelines/date-and-time/).

## React Components

### DateTime

Displays an absolute date and/or time in a variety of formats that conforms to Forma 36 guidelines.

| Format Code    | Example                  |
| -------------- | ------------------------ |
| `FULL`         | `12 Aug 2020 at 8:00 am` |
| `DATE_ONLY`    | `12 Aug 2020`            |
| `TIME_ONLY`    | `8:00 am`                |
| `WEEKDAY_DATE` | `Wed, 12 Aug`            |

**Example Usage**

```ts
import React from 'react';
import { DateTime } from '@contentful/forma36-react-components';

export const MyComponent: React.FC<{ date: Date }> = ({ date }) => {
  return <DateTime date={date} />;
};
```

or with a different format

```ts
import React from 'react';
import { DateTime } from '@contentful/forma36-react-components';

export const ScheduledForWeekdate: React.FC<{ date: Date }> = ({ date }) => {
  return <DateTime date={date} format="WEEKDAY_DATE" />;
};
```

### RelativeDate

Provides a relative date string for resolutions ranging from seconds up through years. Should be used when absolute date and time is unnecessarily specific.

**NOTE**: The component includes a `setInterval` that will automatically trigger a re-render on a per-second basis to keep the value accurate over time.

**Example Usage**

```ts
import React from 'react';
import { RelativeDate } from '@contentful/forma36-react-components';

export const EntryPublishedRelative: React.FC<{ publishedAt: Date }> = ({
  publishedAt,
}) => {
  return <RelativeDate date={publishedAt} />;
};
```

With a different `baseDate`

```ts
import React from 'react';
import { RelativeDate } from '@contentful/forma36-react-components';

export const EntryLifetime: React.FC<{
  createdAt: Date;
  publishedAt: Date;
}> = ({ createdAt, publishedAt }) => {
  return <RelativeDate date={publishedAt} baseDate={createdAt} />;
};
```

## Utility Functions

Additional functions are available for strings in cases when the date and/or time needs to be embedded in plaintext.

| Method                   | Description                                                                                     |
| ------------------------ | ----------------------------------------------------------------------------------------------- |
| `formatDateTime`         | Given a JS date and an optional format returns a string matching the format. Defaults to `FULL` |
| `formatDate`             | Given a JS date returns only the date portion as a string                                       |
| `formatTime`             | Given a JS date returns only the time portion as a string                                       |
| `formatWeekdayDate`      | Given a JS date returns the date portion prefixed with the 3 letter weekday                     |
| `formatRelativeDateTime` | Given a JS date and an optional `baseDate` calculates the relative time difference              |

**NOTE** `formatRelativeDateTime` returns a static string, meaning you will need to trigger a re-render if the relative time is to stay accurate for near-term times. `<RelativeDate />` does this for you automatically.

**Example**

```ts
import { datetime } from '@contentful/forma36-react-components';

const publishedTime = datetime.formatTime(date);
```
