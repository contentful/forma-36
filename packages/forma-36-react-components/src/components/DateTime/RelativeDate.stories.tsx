import React from 'react';
import { storiesOf } from '@storybook/react';
import { text, date } from '@storybook/addon-knobs';

import RelativeDate from './RelativeDate';

const THREE_DAYS_MS = 3 * 24 * 60 * 60 * 1000;
const defaultDate = new Date(Date.now() - THREE_DAYS_MS);
const futureDate = new Date(Date.now() + THREE_DAYS_MS);

storiesOf('Components/RelativeDate', module)
  .addParameters({
    propTypes: RelativeDate['__docgenInfo'],
    component: RelativeDate,
  })
  .add('default', () => (
    <RelativeDate
      date={new Date(date('date', defaultDate))}
      testId={text('testId', '')}
      className={text('className', '')}
    />
  ))
  .add('with a day in the future', () => <RelativeDate date={futureDate} />)
  .add('with a different comparison date', () => (
    <RelativeDate
      date={new Date(date('date', defaultDate))}
      baseDate={new Date(date('baseDate', futureDate))}
    />
  ));
