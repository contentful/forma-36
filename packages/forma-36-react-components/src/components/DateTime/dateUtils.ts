import dayjs from 'dayjs';
import RelativeTime from 'dayjs/plugin/relativeTime';
import { CoercibleDate } from './types';
dayjs.extend(RelativeTime);

const formatTokens = {
  FULL: 'D MMM YYYY[ at ]h:mm a',
  DATE_ONLY: 'D MMM YYYY',
  TIME_ONLY: 'h:mm a',
  WEEKDAY_DATE: 'ddd[, ]D MMM',
};

export type DateTimeFormat = keyof typeof formatTokens;

/**
 * @example
 * > formatDateTime(date)
 * 13 Aug 2019 at 10:00 am
 *
 * @example
 * > formatDateTime(date, 'DATE_ONLY')
 * 13 Aug 2019
 */
export const formatDateTime = (
  date: CoercibleDate,
  token: DateTimeFormat = 'FULL',
): string => {
  if (!formatTokens[token]) {
    throw new TypeError(`Unknown date format '${token}'`);
  }
  return dayjs(date).format(formatTokens[token]);
};

/**
 * @example
 * > formatDate(date)
 * 13 Aug 2019
 */
export const formatDate = (date: CoercibleDate): string => {
  return formatDateTime(date, 'DATE_ONLY');
};

/**
 * @example
 * > formatTime(date)
 * 8:00 am
 */
export const formatTime = (date: CoercibleDate): string => {
  return formatDateTime(date, 'TIME_ONLY');
};

/**
 * @example
 * > formatWeekdayDate(date)
 * Mon, 12 Aug
 */
export const formatWeekdayDate = (date: CoercibleDate): string => {
  return formatDateTime(date, 'WEEKDAY_DATE');
};

export const formatRelativeDateTime = (
  date: CoercibleDate,
  baseDate: CoercibleDate = new Date(),
): string => {
  return dayjs(date).from(baseDate);
};
