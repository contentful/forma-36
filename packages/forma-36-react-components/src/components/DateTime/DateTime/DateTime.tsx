import React from 'react';
import { DateTimeFormat, formatDateTime } from '../dateUtils';
import { CoercibleDate } from '../types';

export interface DateTimeProps {
  /**
   * The date to display as a JS Date, an ISO8601 Timestamp, or Unix Epoch Milliseconds
   *
   * @example '2020-08-13T01:23:45.000Z'
   * @example new Date()
   * @example Date.now()
   */
  date: CoercibleDate;
  /**
   * Which display format to use
   *
   * @default 'FULL'
   **/
  format?: DateTimeFormat;
  className?: string;
  testId?: string;
}

/**
 * Formats a JS Date object to an absolute date and time inside of a <time /> tag
 */
export const DateTime: React.FC<DateTimeProps> = ({
  date,
  format = 'FULL',
  className,
  testId,
}) => {
  if (typeof date === 'string' || typeof date === 'number') {
    date = new Date(date);
  }
  const formatted = formatDateTime(date, format);

  return (
    <time
      dateTime={date.toISOString()}
      className={className}
      data-test-id={testId}
    >
      {formatted}
    </time>
  );
};

export default DateTime;
