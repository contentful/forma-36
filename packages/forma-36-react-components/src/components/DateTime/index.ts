import { DateTime } from './DateTime/DateTime';
import { RelativeDate } from './RelativeDate/RelativeDate';
import * as datetime from './dateUtils';
export { datetime, DateTime, RelativeDate };
