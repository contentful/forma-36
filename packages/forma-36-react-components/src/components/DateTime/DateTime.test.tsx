import React from 'react';
import { render } from '@testing-library/react';

import { DateTime } from './DateTime';

const exampleDate = new Date('2020-04-09T16:17:18.912Z');

const silenceErrorsWithMessage = (errorMessage: string) => {
  // Throw expected error silently
  const originalFunction = console.error;
  const spy = jest.spyOn(console, 'error');
  spy.mockImplementation((error) => {
    if (error.toString().includes(errorMessage)) {
      return;
    } else if (error.toString().includes('The above error occurred')) {
      // Also block subsequent error
      return;
    }
    originalFunction(error);
  });
  return () => spy.mockRestore();
};

describe('DateTime', () => {
  it('renders the component', () => {
    const { container } = render(
      <DateTime
        date={exampleDate}
        className="custom--className"
        testId="custom-testId"
      />,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('allows for FULL formatting', () => {
    const { container } = render(<DateTime date={exampleDate} format="FULL" />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('allows for WEEKDAY_DATE formatting', () => {
    const { container } = render(
      <DateTime date={exampleDate} format="WEEKDAY_DATE" />,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('allows for DATE_ONLY formatting', () => {
    const { container } = render(
      <DateTime date={exampleDate} format="DATE_ONLY" />,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('allows for TIME_ONLY formatting', () => {
    const { container } = render(
      <DateTime date={exampleDate} format="TIME_ONLY" />,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('does not allow for an unknown format', () => {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const format = 'NOT_REAL' as any;
    const restoreErrorPatch = silenceErrorsWithMessage(
      "Unknown date format 'NOT_REAL'",
    );
    expect(() => {
      render(<DateTime date={exampleDate} format={format} />);
    }).toThrow(`Unknown date format 'NOT_REAL'`);
    restoreErrorPatch();
  });
});
