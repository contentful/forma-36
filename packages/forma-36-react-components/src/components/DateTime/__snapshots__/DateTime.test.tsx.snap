// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateTime allows for DATE_ONLY formatting 1`] = `
<time
  datetime="2020-04-09"
>
  9 Apr 2020
</time>
`;

exports[`DateTime allows for FULL formatting 1`] = `
<time
  datetime="2020-04-09T16:17:18.912Z"
>
  9 Apr 2020 at 9:17 am
</time>
`;

exports[`DateTime allows for TIME_ONLY formatting 1`] = `
<time
  datetime="16:17:18.912"
>
  9:17 am
</time>
`;

exports[`DateTime allows for WEEKDAY_DATE formatting 1`] = `
<time
  datetime="04-09"
>
  Thu, 9 Apr
</time>
`;

exports[`DateTime renders the component 1`] = `
<time
  class="custom--className"
  data-test-id="custom-testId"
  datetime="2020-04-09T16:17:18.912Z"
>
  9 Apr 2020 at 9:17 am
</time>
`;
